### PR TITLE
Use type inference instead of multiple declarations

### PR DESCRIPTION
**Pull request by koshic**
See #580.

Change types from

```ts
export function execa(file: string, options?: Options): ExecaChildProcess;
export function execa(file: string, options?: Options<BufferEncodingOption>): ExecaChildProcess<Buffer>;
```

to

```ts
export function execa<EncodingType extends EncodingOption = DefaultEncodingOption>(
	file: string, options?: Options<EncodingType>
): ExecaChildProcess<EncodingType extends DefaultEncodingOption ? string : Buffer>;
```

It's a prerequisite to make smarter return type which allow us to properly infer `pipeXxx` types.

### CODE REVIEW

**Comment by koshic** [comment]
@ehmicky could you plz take a look? there is a failed test, but it doesn't related to .d.ts changes

File: index.d.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ type EncodingOption =
 type DefaultEncodingOption = 'utf8';
 type BufferEncodingOption = 'buffer' | null;
 
-export type CommonOptions<EncodingType extends EncodingOption = DefaultEncodingOption> = {
```
Comment (ehmicky) [inline index.d.ts:35]: Why remove all the default values for the generic types? This makes it easier to use the types for the common use case (where `Buffer`s are not used).

Comment (koshic) [reply index.d.ts:35]: It helps to be ensured that all variants of execa definition will have explicit type parameter. But I've forget about users who may use exported types in their apps, my bad. 

I'll restore those defaults, thx!

File: index.d.ts [hunk 2]
```diff
@@ -637,18 +637,14 @@ setTimeout(() => {
...
-): ExecaChildProcess;
-export function execa(
-	file: string,
-	arguments?: readonly string[],
-	options?: Options<BufferEncodingOption>
-): ExecaChildProcess<Buffer>;
-export function execa(file: string, options?: Options): ExecaChildProcess;
-export function execa(file: string, options?: Options<BufferEncodingOption>): ExecaChildProcess<Buffer>;
+	options?: Options<EncodingType>
+): ExecaChildProcess<EncodingType extends DefaultEncodingOption ? string : Buffer>;
```
Comment (ehmicky) [inline index.d.ts:644]: Should we create a type for `EncodingType extends DefaultEncodingOption ? string : Buffer` since it is used multiple times?

Comment (koshic) [reply index.d.ts:644]: I thought about that but decided to keep that PR simple. At the next step we already should create complicated type set to infer results. But if you want, I can add something like `GetStdoutStderrType<EncodingType extends EncodingOption>`

Comment (ehmicky) [reply index.d.ts:644]: > But if you want, I can add something like GetStdoutStderrType<EncodingType extends EncodingOption>

Yes, that sounds good. I think using such a type would actually simplify the PR, by making it shorter and with fewer `extends ? :` statements.

Comment (koshic) [reply index.d.ts:644]: Ok, it shouldn't take too much time. But plz be ready to re-run those unstable tests )

**Review by ehmicky — APPROVED** [review]
This is looking good to me. Thanks a lot @koshic!

What do you think @sindresorhus?

**Pull request merged**